java-checks/src/main/java/org/sonar/java/checks/security/SMTPSSLServerIdentityCheck.java

```diff
@@ -0,0 +1,150 @@
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ConstantUtils;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.TypeCriteria;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S4499")
+public class SMTPSSLServerIdentityCheck extends AbstractMethodDetection {
+
+  private static final String EMAIL_CLASS_NAME = "org.apache.commons.mail.Email";
+  private static final String SESSION_CLASS_NAME = "javax.mail.Session";
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      MethodMatcher.create()
+        .typeDefinition(TypeCriteria.is(EMAIL_CLASS_NAME))
+        .name("send")
+        .withoutParameter(),
+      MethodMatcher.create()
+        .typeDefinition(TypeCriteria.is(SESSION_CLASS_NAME))
+        .name("getDefaultInstance")
+        .withAnyParameters());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    MethodTree method = findEnclosingMethod(mit);
+    MethodBodyVisitor methodVisitor = null;
+    if (method != null) {
+      if ("send".equals(mit.symbol().name())) {
+        methodVisitor = new MethodBodyVisitor(0);
+      } else {
+        methodVisitor = new MethodBodyVisitor(1);
+      }
+      method.accept(methodVisitor);
+      if (methodVisitor.bothFunctionsAreChecked < 2) {
+        if (methodVisitor.differentiateEmailAndSessionInvocations == 0) {
+          reportIssue(mit.methodSelect(), "Enable server identity validation on this SMTP SSL connection.");
+        } else if (methodVisitor.differentiateEmailAndSessionInvocations == 1) {
+          reportIssue(mit.methodSelect(), "Enable server identity validation, set \"mail.smtp.ssl.checkserveridentity\" to true");
+        }
+      }
+    }
+    super.onMethodInvocationFound(mit);
+  }
+
+  @CheckForNull
+  private static MethodTree findEnclosingMethod(Tree tree) {
+    while (!tree.is(Tree.Kind.CLASS, Tree.Kind.METHOD)) {
+      tree = tree.parent();
+    }
+    if (tree.is(Tree.Kind.CLASS)) {
+      return null;
+    }
+    return (MethodTree) tree;
+  }
+
+  private static class MethodBodyVisitor extends BaseTreeVisitor {
+
+    private Integer bothFunctionsAreChecked = 0;
+    /*
+     * MethodVisitor constructor's parameter is 0 when "send()" from org.apache.commons.mail.Email is invoked
+     * and 1 for "getDefaultInstance" from javax.mail.Session
+     */
+    private final Integer differentiateEmailAndSessionInvocations;
```

Comment:
This field is used to trigger 2 different behaviors. We should have 2 different classes instead.